website/blog/2023-03-08-mongodb-vs-postgresql.md

```diff
@@ -0,0 +1,163 @@
+---
+slug: mongodb-vs-postgresql-database-comparison
+title: "MongoDB vs PostgreSQL: A Detailed Database Comparison"
+authors: [alex]
+description: >
+    Compare the features and benefits of MongoDB and PostgreSQL to determine which database management system is best for your application.
+image: /img/blog/mongodb-postgres.jpg
+keywords: [MongoDB, PostgreSQL, NoSQL, relational databases, scalability, performance, data modeling, schema design]
+tags: [document database, mongodb alternative, mongodb compatible]
+---
+
+![MongoDB vs PostgreSQL](/img/blog/mongodb-postgres.jpg)
+
+Everyone has been there – starting a new project and wondering what database to use.
+Deciding between MongoDB and PostgreSQL?
+NoSQL vs SQL database?
+Open source vs Proprietary database?
+Surely there should be one that suits your project to a tee.
+
+<!--truncate-->
+
+According to a [2022 Statista survey](https://www.statista.com/statistics/809750/worldwide-popularity-ranking-database-management-systems/), PostgreSQL and MongoDB are two of the most popular databases today, ranking 4th and 5th, respectively.
+Both databases share a few similarities and striking differences in terms of architecture, data model, commands, use cases, security, and more.
+
+This article will explore both PostgreSQL and MongoDB databases in detail and assist you in making an informed decision.
+
+Let's get started.
+
+## Overview of MongoDB and PostgreSQL
+
+### MongoDB
+
+[MongoDB](https://www.mongodb.com/) is a user-friendly, schema-free, proprietary NoSQL database written as JSON-like (BSON) documents for storing data.
+It is a document database that uses BSON (stands for binary JSON), which offers more data types than regular JSON data, such as floating point, date, etc.
+See [how we convert and store MongoDB BSON to PostgreSQL JSON](https://blog.ferretdb.io/pjson-how-to-store-bson-in-jsonb/).
+
+Unlike most traditional databases using SQL, MongoDB uses a different syntax and structure, which is relatively easy to learn, even for non-programmers.
+NoSQL databases like MongoDB are well suited for managing semi-structured or unstructured data.
+You can easily add new fields or data types without explicitly declaring the document's structure, making it more flexible for users to view the data, edit it, or update the schema as much as they want.
+This flexibility also enables MongoDB to process and handle large amounts of data faster than many other databases.
```

Comment:
I would go into some details about how flexibility affects the performance, because currently it looks like the fact that user can structure data as they want, makes handling large amounts of data faster. 

We could mention that SQL databases need to do much more operations before inserting the data (like type checking on the server side or schema validation), while MongoDB rather puts this data as they were specified on the client side. 
Obtaining data is also different as for example doing JOINs on every operation on tables with relations, on SQL databases makes query evaluation slower. 

Overall, we could say a few things about joins in this article. As PostgreSQL offers referring in the row of one table to other rows from another tables. It allows users to create relational structure, that can return structured data from multiple tables via a single JOIN query.

However as MongoDB does not support relations, JOINs are not a thing there. So the approach is to put all necessary data (like shopping cart list) of user directly inside it's object, which makes a structure less organized, but queries should be faster (as DBMS to query the data, doesn't need to evaluate relations from different places).

It also should be easier to scale database horizontally (as all required data for query is in one place, you don't need to invent any complex solution to handle relations from sharded tables across multiple machines).

Reply:
Many good points. PostgreSQL has vast amount of features but fairly complex. Even things that seem straightforward like changing a column type can be slow and complex if you have millions of entries. With mongodb, the flexibility of schema less structure allows us to not have that problem.